### PR TITLE
[time-tests] Remove outliers using Inter Quartile Range (IQR)

### PR DESCRIPTION
**Pull request by just-sparta**
- *51329

### CODE REVIEW

**Review by asomsiko — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/openvinotoolkit/openvino/5014)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @just-sparta and @vurusovs)

___
*[tests/time_tests/scripts/run_timetest.py, line 24 at r1](https://reviewable.io/reviews/openvinotoolkit/openvino/5014#-MX1OosQ4eBNYgi_q8tJ:-MX1OosR8ANbPE1mqt_P:b1it0ot) ([raw file](https://github.com/openvinotoolkit/openvino/blob/19ee65f8696a2fbb296baf0895d49677cce5e8c9/tests/time_tests/scripts/run_timetest.py#L24)):*
> ```Python
> from pathlib import Path
> from pprint import pprint
> import numpy as np
> ```

please update requirements.txt with numpy
___
*[tests/time_tests/scripts/run_timetest.py, line 111 at r1](https://reviewable.io/reviews/openvinotoolkit/openvino/5014#-MX1QDdRDlrFHySddgq1:-MX1QDdRDlrFHySddgq2:b-oq4t2d) ([raw file](https://github.com/openvinotoolkit/openvino/blob/19ee65f8696a2fbb296baf0895d49677cce5e8c9/tests/time_tests/scripts/run_timetest.py#L111)):*
> ```
> 1.5
> ```
Please define a constant for this magic number. E.g.

```
# Define a range to cut outliers which are < Q1 − IRQ_CUTOFF * IQR, and > Q3 + IRQ_CUTOFF * IQR
# https://en.wikipedia.org/wiki/Interquartile_range
IRQ_CUTOFF = 1.5
```

**Review by asomsiko — COMMENTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/openvinotoolkit/openvino/5014)* status: all files reviewed, 2 unresolved discussions (waiting on @just-sparta and @vurusovs)

File: tests/time_tests/scripts/run_timetest.py [hunk 1]
```diff
@@ -95,6 +109,13 @@ def run_timetest(args: dict, log=None):
         stats = dict((step_name, stats.get(step_name, []) + [duration])
                      for step_name, duration in raw_data.items())
 
+    # Remove outliers
+    for step_name, time_results in stats.items():
+        iqr, q1, q3 = calculate_iqr(time_results)
```
Comment (vurusovs) [inline tests/time_tests/scripts/run_timetest.py:114]: @just-sparta may me move all calculation and filtering to a utilities to decrease number fo changes in `run_timetest.py`?

File: tests/time_tests/scripts/run_timetest.py [hunk 2]
```diff
@@ -11,16 +11,22 @@
...
-from pprint import pprint
 import yaml
 
+from pathlib import Path
+from pprint import pprint
+import numpy as np
+
+# Define a range to cut outliers which are < Q1 − IRQ_CUTOFF * IQR, and > Q3 + IRQ_CUTOFF * IQR
+# https://en.wikipedia.org/wiki/Interquartile_range
+IRQ_CUTOFF = 1.5
```
Comment (vurusovs) [inline tests/time_tests/scripts/run_timetest.py:28]: Looks like a missprint for `IQR_CUTOFF`

**Review by asomsiko — COMMENTED** [review]
> Reviewed 1 of 3 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/openvinotoolkit/openvino/5014)* status: all files reviewed, 3 unresolved discussions (waiting on @just-sparta)

___
*[tests/time_tests/test_runner/conftest.py, line 60 at r3](https://reviewable.io/reviews/openvinotoolkit/openvino/5014#-MX5pstVCVUFp2Oshbw2:-MX5pstVCVUFp2Oshbw3:b-wrpjbh) ([raw file](https://github.com/openvinotoolkit/openvino/blob/f5910a27f626e374cc64fb7586ee8e733d3ad2c1/tests/time_tests/test_runner/conftest.py#L60)):*
> ```Python
>         type=check_positive_int,
>         help="number of iterations to run executable and aggregate results",
>         default=3
> ```

@just-sparta,  I thought we wanted to increase the number of iterations after introducing a cutoff. You did it on r1 why you have rolled it back to 3 now in r1?

**Review by asomsiko — COMMENTED** [review]
> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/openvinotoolkit/openvino/5014)* status: all files reviewed, 3 unresolved discussions (waiting on @just-sparta)

**Comment by just-sparta** [comment]
> > Reviewed 1 of 3 files at r2, 3 of 3 files at r3.
> > _[Reviewable](https://reviewable.io/reviews/openvinotoolkit/openvino/5014)_ status: all files reviewed, 3 unresolved discussions (waiting on @just-sparta)
> 
> _[tests/time_tests/test_runner/conftest.py, line 60 at r3](https://reviewable.io/reviews/openvinotoolkit/openvino/5014#-MX5pstVCVUFp2Oshbw2:-MX5pstVCVUFp2Oshbw3:b-wrpjbh) ([raw file](https://github.com/openvinotoolkit/openvino/blob/f5910a27f626e374cc64fb7586ee8e733d3ad2c1/tests/time_tests/test_runner/conftest.py#L60)):_
> 
> > ```python
> >         type=check_positive_int,
> >         help="number of iterations to run executable and aggregate results",
> >         default=3
> > ```
> 
> @just-sparta, I thought we wanted to increase the number of iterations after introducing a cutoff. You did it on r1 why you have rolled it back to 3 now in r1?

@asomsiko it is necessary to implement a smart calculation of iterations for models depending on the execution time. WE can increase the number of iterations for models that run for a small amount of time

**Review by asomsiko — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/openvinotoolkit/openvino/5014)* status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @asomsiko and @just-sparta)

___
*[tests/time_tests/test_runner/conftest.py, line 60 at r3](https://reviewable.io/reviews/openvinotoolkit/openvino/5014#-MX5pstVCVUFp2Oshbw2:-MX7JeOvEiXHfdGddxCA:b218a3e) ([raw file](https://github.com/openvinotoolkit/openvino/blob/f5910a27f626e374cc64fb7586ee8e733d3ad2c1/tests/time_tests/test_runner/conftest.py#L60)):*
> ease the number of iterations for models th
since cutoff is enabled by default lets increase default iterations to 10

**Comment by just-sparta** [comment]
> > _[Reviewable](https://reviewable.io/reviews/openvinotoolkit/openvino/5014)_ status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @asomsiko and @just-sparta)
> 
> _[tests/time_tests/test_runner/conftest.py, line 60 at r3](https://reviewable.io/reviews/openvinotoolkit/openvino/5014#-MX5pstVCVUFp2Oshbw2:-MX7JeOvEiXHfdGddxCA:b218a3e) ([raw file](https://github.com/openvinotoolkit/openvino/blob/f5910a27f626e374cc64fb7586ee8e733d3ad2c1/tests/time_tests/test_runner/conftest.py#L60)):_
> 
> > ease the number of iterations for models th
> > since cutoff is enabled by default lets increase default iterations to 10

iterations value increased to 10

**Comment by just-sparta** [comment]
@asomsiko please resolve if there are no objections

**Review by vurusovs — APPROVED** [review]

**Review by asomsiko — APPROVED** [review]
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/openvinotoolkit/openvino/5014)* status: all files reviewed, 2 unresolved discussions (waiting on @just-sparta)

**Pull request merged**